### PR TITLE
alloc_cpu: Use plain malloc instead of posix_memalign

### PR DESCRIPTION
**Pull request by peterbell10**
This replaces `posix_memalign` with `malloc` plus manual pointer arithmetic to return an aligned array. On my machine this saves around 30-50 ns per allocation over the course of a `timeit` benchmark. Which for small, contiguous arrays is around 3-4% of overall operator call time.

Looking back, [`THAllocInternal`](https://github.com/pytorch/pytorch/blob/a24163a95edb193ff7b06e98cd69bf7cfd4c0d2f/aten/src/TH/THGeneral.cpp#L161) used to use `malloc` for allocations less than 5 KB and `posix_memalign` only for large allocations. So, it's possible this performance difference was known before but forgotten when moving it to `c10`.

### CODE REVIEW

File: c10/core/CPUAllocator.cpp [hunk 1]
```diff
@@ -55,16 +55,19 @@ void* alloc_cpu(size_t nbytes) {
...
+  const size_t align = std::max(gAlignment, sizeof(void*));
+  void* storage = malloc(nbytes + align);
+  if (storage) {
+    auto storage_addr = reinterpret_cast<uintptr_t>(storage);
+    auto data_addr = storage_addr + align;
+    data_addr -= data_addr % align;
+    data = reinterpret_cast<void*>(data_addr);
+    reinterpret_cast<void**>(data_addr)[-1] = storage;
+  } else {
+    data = nullptr;
```
Comment (swolchok) [inline c10/core/CPUAllocator.cpp:70]: why don't we need to throw if allocation fails like we did before?

File: c10/core/CPUAllocator.cpp [hunk 2]
```diff
@@ -55,16 +55,19 @@ void* alloc_cpu(size_t nbytes) {
...
-        "DefaultCPUAllocator: can't allocate memory: you tried to allocate ",
-        nbytes,
-        " bytes. Error code ",
-        err,
-        " (",
-        strerror(err),
-        ")");
+  // Allocate storage with malloc and manually align data pointer
+  // Then stash storage starting address before the first element to free later
+  // This is significantly faster than posix_memalign
```
Comment (swolchok) [inline c10/core/CPUAllocator.cpp:60]: I imagine that this is dependent on the operating system and allocator in use; can you clarify where it was measured? In particular, I would expect jemalloc to do a decent job with posix_memalign...

**Review by swolchok — CHANGES_REQUESTED** [review]
I'm not sure about this one. Should we use the C++17 aligned_alloc if available? I'll see who else I can find to chime in on this...

**Comment by swolchok** [comment]
Added @dzhulgakov, @smessmer, and @ezyang because I'm not sure about this one.

File: c10/core/CPUAllocator.cpp (hunk 2 above)
Comment (peterbell10) [reply c10/core/CPUAllocator.cpp:60]: This was measured on ubuntu linux with default malloc.

File: c10/core/CPUAllocator.cpp (hunk 1 above)
Comment (peterbell10) [reply c10/core/CPUAllocator.cpp:70]: The `CAFFE_ENFORCE` below raises when `data` is `nullptr`.

**Comment by ezyang** [comment]
Not commenting on the PR as a whole as I haven't read it carefully yet, but I'm pretty sure what happened is someone convinced us that `posix_memalign` was just as good as `malloc`, so it was OK to use it unconditionally. I'll have to dig up the notes.

**Comment by ezyang** [comment]
Found it: https://github.com/pytorch/pytorch/pull/14517#discussion_r237726929

File: c10/core/CPUAllocator.cpp [hunk 3]
```diff
@@ -55,16 +55,19 @@ void* alloc_cpu(size_t nbytes) {
...
+  const size_t align = std::max(gAlignment, sizeof(void*));
```
Comment (ezyang) [inline c10/core/CPUAllocator.cpp:61]: FYI: gAlignment is 64 bytes on server!

**Comment by ezyang** [comment]
cc @colesbury

I'm not sure we necessarily want to take the time-space tradeoff here. You're getting this speedup by unconditionally overallocating an extra 64 bytes for every tensor, and given this comment I'm not sure we actually want this.

```
// Note [TensorImpl size constraints]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Changed the size of TensorImpl?  If the size went down, good for
// you!  Adjust the documentation below and the expected size.
// Did it go up?  Read on...
//
// Struct size matters.  In some production systems at Facebook, we have
// 400M live tensors during a training run.  Do the math: every 64-bit
// word you add to Tensor is an extra 3.2 gigabytes in RAM.
//
// If you are a Facebook employee, you can check if the run in question
// has tipped you over the point using the command here:
// https://fburl.com/q5enpv98
//
// For reference, we OOMed at 160 bytes (20 words) per TensorImpl.
// This is not counting overhead from strides out-of-line allocation and
// StorageImpl space and this is from before we inlined sizes and strides
// directly into TensorImpl as SmallVectors.
//
// Our memory usage on 32-bit systems is suboptimal, but we're not checking
// for it at the moment (to help avoid rage inducing cycles when the
// 32-bit number is wrong).
```

**Comment by peterbell10** [comment]
>Found it: https://github.com/pytorch/pytorch/pull/14517#discussion_r237726929

It seems the benchmarks were never done?

> You're getting this speedup by unconditionally overallocating an extra 64 bytes for every tensor,

To be precise, it's an extra 64 bytes for every `Storage`. That's a non-trivial distinction since there can be many tensors sharing the same storage. However, I admit I don't know how relevant that is to FB's use case.

To maximize benefit without too much memory overhead, perhaps using different strategies based on allocation size (similar to `THAlloc`) would make sense: 
- For small allocations, I would suggest it doesn't make sense to align any more than the next power of 2. For example, aligning a scalar tensor to 64 bytes doesn't have any benefit in terms of memory access or vectorization. It's just a waste of memory.
- For large tensors (say 4KB), allocation time should be small compared to computation time so we could use `posix_memalign` there.

Does that sound more reasonable?

**Comment by ezyang** [comment]
I asked @colesbury for his thoughts and he sent this in:

> 1) I don't love the manual alignment strategy because it's wasteful when you use something like jemalloc, and can make debugging harder (you don't have a pointer to the beginning of the allocation)
> 2) Peter's suggestion of using posix_memalign only for larger allocations makes sense
> 3) We should "vendor" jemalloc and use it internally for the allocation of Tensor's storages. There will be additional perf gains compared to glibc's malloc especially for large tensors. That's a bigger change than this PR of course, but I think it should be on PyTorch's roadmap.

I agree with all the thoughts here. Are you interested in vendoring jemalloc? I think that's probably a better way to solve the malloc problem.

**Comment by peterbell10** [comment]
Vendoring `jemalloc` sounds interesting. I'll look into it.

**Pull request closed without merging**